libcache/libcache_utils.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Phoenix-RTOS
+ *
+ * phoenix-rtos-tests
+ *
+ * Libcache tests utilities
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Malgorzata Wrobel
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "libcache_utils.h"
+#include "unity_fixture.h"
+
+#include <stdlib.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <time.h>
+#include <unistd.h>
+
+
+void test_genCharFile(void)
+{
+	FILE *file;
+	int i;
+	char num;
+
+	file = fopen("/var/read.txt", "w");
+	fseek(file, 0, SEEK_SET);
+
+	for (i = 0; i < LIBCACHE_SRC_MEM_SIZE; ++i) {
+		num = rand();
+		fwrite_unlocked(&num, sizeof(char), 1, file);
+	}
+
+	fclose(file);
+}
+
+void test_genIntFile(void)
+{
+	FILE *file;
+	int i;
+	int num;
+
+	file = fopen("/var/read.txt", "w");
+	fseek(file, 0, SEEK_SET);
+
+	for (i = 0; i < LIBCACHE_SRC_MEM_SIZE / 4; ++i) {
+		num = rand();
+		fwrite_unlocked(&num, sizeof(int), 1, file);
+	}
+
+	fclose(file);
+}
+
+ssize_t test_readCb(uint64_t offset, void *buffer, size_t count, cache_devCtx_t *ctx)
+{
+	ssize_t ret = -1;
+
+	lseek(srcMem, offset, SEEK_SET);
+	ret = read(srcMem, buffer, count);
+
+	return ret;
+}
+
+ssize_t test_writeCb(uint64_t offset, const void *buffer, size_t count, cache_devCtx_t *ctx)
+{
+	ssize_t ret = -1;
+
+	lseek(srcMem, offset, SEEK_SET);
+	ret = write(srcMem, buffer, count);
+
+	return ret;
+}
+
+ssize_t test_readCbErr(uint64_t offset, void *buffer, size_t count, cache_devCtx_t *ctx)
+{
+	return -EIO;
+}
+
+ssize_t test_writeCbErr(uint64_t offset, const void *buffer, size_t count, cache_devCtx_t *ctx)
+{
+	return -EIO;
+}
+
+void *test_cache_write(void *args)
+{
+	ssize_t written = -1;
+	test_write_args_t *arguments = args;
+
+	written = cache_write(arguments->cache, arguments->addr, arguments->buffer, arguments->count, arguments->policy);
+	TEST_ASSERT_EQUAL_INT(arguments->count, written);
```

Comment:
Please note that Unity asserts should only be used in the main thread context. It's a macro for longjmp therefore when the test fails some weird things can happen. Threads could communicate by shared state and the main thread eventually can verify it using assertions. (similar to https://github.com/phoenix-rtos/phoenix-rtos-tests/pull/121)

Reply:
Thank you for pointing that out. This assertion is important from the point of view of the test though, as I need to test the number of bytes written by `cache_write` function.

I've figured out a workaround: I capture the value returned from the thread function by passing a pointer to a buffer to obtain that value. Unfortunately this approach causes an overhead of a few additional `malloc` and `free` calls.

A possible solution without `malloc` would be adding an additional field to `test_write_args_t` to capture the number of written bytes so that I could make an assertion `TEST_ASSERT_EQUAL_INT(args->written, ...) in the main thread context before proceeding.

I think that the first solution is better as it simply makes use of the `**__thread_return` argument of the `int pthread_join(pthread_t __th, void **__thread_return)` function.